internal/controller/dnsrecord_controller.go

```diff
@@ -257,42 +296,57 @@ func (r *DNSRecordReconciler) publishRecord(ctx context.Context, dnsRecord *v1al
 	}
 	err := r.Get(ctx, client.ObjectKeyFromObject(managedZone), managedZone, &client.GetOptions{})
 	if err != nil {
-		return noRequeueDuration, err
+		return false, err
 	}
 	managedZoneReady := meta.IsStatusConditionTrue(managedZone.Status.Conditions, "Ready")
 
 	if !managedZoneReady {
-		return noRequeueDuration, fmt.Errorf("the managed zone is not in a ready state : %s", managedZone.Name)
+		return false, fmt.Errorf("the managed zone is not in a ready state : %s", managedZone.Name)
 	}
 
-	// cut off here for the short reconcile loop
-	requeueIn := validFor
-	if dnsRecord.Status.ValidFor != "" {
-		requeueIn, _ = time.ParseDuration(dnsRecord.Status.ValidFor)
-	}
-	expiryTime := metav1.NewTime(dnsRecord.Status.QueuedAt.Add(requeueIn))
-	if !generationChanged(dnsRecord) && reconcileStart.Before(&expiryTime) {
+	if prematurely, _ := recordReceivedPrematurely(dnsRecord); prematurely {
 		logger.V(1).Info("Skipping managed zone to which the DNS dnsRecord is already published and is still valid", "dnsRecord", dnsRecord.Name, "managedZone", managedZone.Name)
-		return requeueIn, nil
-	}
-	if generationChanged(dnsRecord) {
-		dnsRecord.Status.WriteCounter = 0
-		wrtiteCounter.WithLabelValues(dnsRecord.Name, dnsRecord.Namespace).Set(0)
+		return false, nil
 	}
 
-	requeueAfter, err := r.applyChanges(ctx, dnsRecord, managedZone, false)
+	hadChanges, err := r.applyChanges(ctx, dnsRecord, managedZone, false)
 	if err != nil {
-		return noRequeueDuration, err
+		return hadChanges, err
 	}
 	logger.Info("Published DNSRecord to manage zone", "dnsRecord", dnsRecord.Name, "managedZone", managedZone.Name)
 
-	return requeueAfter, nil
+	return hadChanges, nil
+}
+
+func recordReceivedPrematurely(record *v1alpha1.DNSRecord) (bool, time.Duration) {
+	requeueIn := validFor
+	if record.Status.ValidFor != "" {
+		requeueIn, _ = time.ParseDuration(record.Status.ValidFor)
+	}
+	expiryTime := metav1.NewTime(record.Status.QueuedAt.Add(requeueIn))
+	return !generationChanged(record) && reconcileStart.Before(&expiryTime), requeueIn
 }
```

Comment:
What does `ReceivedPrematurely` mean? Can you add a description to the function?

Reply:
Added a comment. Hopefully, it will make it more obvious. What it does can be summarized as `return reconcileStart.Before(lastReconcile.Add(validFor))`